src/ovro_lwa_portal/io.py

```diff
@@ -0,0 +1,373 @@
+
+    except ImportError:
+        # Fallback to DOI.org resolution
+        logger.info(f"caltechdata_api not available, using DOI.org resolution for: {doi}")
+        doi_url = f"https://doi.org/{doi}"
+
+        try:
+            response = requests.get(
+                doi_url,
+                headers={"Accept": "application/json"},
+                allow_redirects=True,
+                timeout=10,
+            )
+            response.raise_for_status()
+
+            # The final URL after redirects might be the data location
+            final_url = response.url
+            logger.info(f"Resolved DOI to URL: {final_url}")
+            return final_url
+
+        except requests.RequestException as e:
+            msg = f"Failed to resolve DOI {doi}: {e}"
+            raise DataSourceError(msg) from e
+
+
+def _detect_source_type(source: str | Path) -> tuple[str, str]:
+    """Detect the type of data source.
+
+    Parameters
+    ----------
+    source : str or Path
+        Data source to analyze.
+
+    Returns
+    -------
+    tuple[str, str]
+        Tuple of (source_type, normalized_source) where source_type is one of:
+        'local', 'remote', 'doi'.
+    """
+    source_str = str(source)
+
+    # Check for DOI
+    if _is_doi(source_str):
+        normalized_doi = _normalize_doi(source_str)
+        return ("doi", normalized_doi)
+
+    # Check for remote URL
+    parsed = urlparse(source_str)
+    if parsed.scheme in ("http", "https", "s3", "gs", "gcs", "abfs", "az"):
+        return ("remote", source_str)
+
+    # Default to local path
+    return ("local", source_str)
+
+
+def _validate_dataset(ds: xr.Dataset) -> None:
+    """Validate that dataset conforms to OVRO-LWA data model.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset to validate.
+
+    Raises
+    ------
+    DataSourceError
+        If dataset doesn't conform to expected structure.
+    """
+    # Check for required dimensions (at least some of these should exist)
+    expected_dims = {"time", "frequency", "l", "m"}
+    found_dims = set(ds.sizes.keys())
+
+    if not found_dims.intersection(expected_dims):
+        logger.warning(
+            f"Dataset may not be OVRO-LWA format. "
+            f"Expected dimensions like {expected_dims}, found {found_dims}"
+        )
+
+    # Check for common OVRO-LWA data variables
+    common_vars = {"SKY", "BEAM"}
+    found_vars = set(ds.data_vars.keys())
+
+    if not found_vars.intersection(common_vars):
+        logger.warning(
+            f"Dataset may not be OVRO-LWA format. "
+            f"Expected variables like {common_vars}, found {found_vars}"
+        )
+
+    logger.info(f"Dataset loaded with dimensions: {dict(ds.sizes)}")
+    logger.info(f"Dataset variables: {list(ds.data_vars.keys())}")
+
+
+def open_dataset(
+    source: str | Path,
+    chunks: dict[str, int] | str | None = "auto",
+    engine: str = "zarr",
+    validate: bool = True,
```

Comment:
For now, let's leave out validation as we don't have an exact dataset format yet.